ddtrace/internal/codeowners.py

```diff
@@ -0,0 +1,191 @@
+        elif ch == "*":
+            if i + 1 < len(pattern_trimmed) and pattern_trimmed[i + 1] == "*":
+                left_anchored = i == 0
+                leading_slash = i > 0 and pattern_trimmed[i - 1] == "/"
+                right_anchored = i + 2 == len(pattern_trimmed)
+                trailing_slash = i + 2 < len(pattern_trimmed) and pattern_trimmed[i + 2] == "/"
+
+                if (left_anchored or leading_slash) and (right_anchored or trailing_slash):
+                    regex += ".*"
+
+                    next(iterator, None)
+                    next(iterator, None)
+                    continue
+            regex += "[^/]*"
+        elif ch == "?":
+            regex += "[^/]"
+        elif ch == "[":
+            in_char_class = True
+            regex += ch
+        elif ch == "]":
+            if in_char_class:
+                regex += ch
+                in_char_class = False
+            else:
+                regex += re.escape(ch)
+        else:
+            regex += re.escape(ch)
+
+    if in_char_class:
+        raise ValueError("unterminated character class in pattern {pattern}".format(pattern=pattern))
+
+    regex += "/" if matches_dir else r"(?:\Z|/)"
+    return re.compile(regex)
+
+
+class Codeowners(object):
+    """Provide interface to parse CODEOWNERS file and match a given path against it."""
+
+    KNOWN_LOCATIONS = (
+        "CODEOWNERS",
+        ".github/CODEOWNERS",
+        "docs/CODEOWNERS",
+        ".gitlab/CODEOWNERS",
+    )
+
+    def __init__(self, path=None, cwd=None):
+        # type: (Optional[str], Optional[str]) -> None
+        """Initialize Codeowners object.
+
+        :param path: path to CODEOWNERS file otherwise try to use any from known locations
+        """
+        path = path or self.location(cwd)
+        if path is not None:
+            self.path = path  # type: str
+        self.patterns = []  # type: List[Tuple[re.Pattern, List[str]]]
+        self.parse()
+
+    def location(self, cwd=None):
+        # type: (Optional[str]) -> Optional[str]
+        """Return the location of the CODEOWNERS file."""
+        cwd = cwd or os.getcwd()
+        for location in self.KNOWN_LOCATIONS:
+            path = os.path.join(cwd, location)
+            if os.path.exists(path):
+                return path
+        raise ValueError("CODEOWNERS file not found")
+
+    def parse(self):
+        # type: () -> None
+        """Parse CODEOWNERS file and store the lines and regexes."""
+        with open_text(self.path) as f:
+            patterns = []
+            for line in f.readlines():
+                line = line.strip()
+                if line == "":
+                    continue
+                # Lines starting with '#' are comments.
+                if line.startswith("#"):
+                    continue
+                if line.startswith("[") and line.endswith("]"):
+                    continue
+                if line.startswith("^[") and line.endswith("]"):
```

Comment:
what do these mean?

Reply:
I can add links to the documentation:
```suggestion
                if line.startswith("[") and line.endswith("]"):
                    # https://docs.gitlab.com/ee/user/project/code_owners.html#organize-code-owners-by-putting-them-into-sections
                    continue
                if line.startswith("^[") and line.endswith("]"):
                    # https://docs.gitlab.com/ee/user/project/code_owners.html#make-a-code-owners-section-optional
```
or simple comments
```suggestion
                if line.startswith("[") and line.endswith("]"):
                    # found a code owners section
                    continue
                if line.startswith("^[") and line.endswith("]"):
                    # found an optional code owners section
```